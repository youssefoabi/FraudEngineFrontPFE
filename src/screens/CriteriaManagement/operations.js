import { all, call, put, select, takeLatest } from 'redux-saga/effects';
import {
  contains,
  filter,
  isEmpty,
  isNil,
  join,
  length,
  map,
  path,
  pathOr,
  pipe,
  prop,
  propEq,
  values,
} from 'ramda';

import {
  DELETE_CRITERIA_ERROR,
  DELETE_CRITERIA_REQUEST,
  DELETE_CRITERIA_SUCCESS,
  FETCH_CRITERIA,
  FETCH_CRITERIA_ERROR,
  SET_CRITERIA,
  SET_CRITERIA_SELECTED,
} from '../../components/CriteriaDataGrid/actions';
import { PENDING_REQUEST, RESET_PENDING_REQUEST } from '../../components/ServerActionBackdrop/actions';
import clientApi from '../../client-api';

export const parseResult = pathOr([], ['data']);

export function* fetchCriteria() {
  try {
    const result = yield call(clientApi().getCriteria);

    const criteria = parseResult(result);

    yield put({ type: SET_CRITERIA, payload: criteria });

    yield put({ type: SET_CRITERIA_SELECTED, payload: [] });
  } catch (error) {
    yield put({ type: FETCH_CRITERIA_ERROR, error });
  }
}

export function* deleteCriterion(id) {
  try {
    yield call(clientApi().deleteCriterion, id);
  } catch (error) {
    const name = yield select(path(['entities', 'criteria', 'byId', id, 'name']));
    return { isSuccess: false, name, id };
  }

  return { isSuccess: true, id };
}

export function* deleteCriteria() {
  yield put({ type: PENDING_REQUEST });

  try {
    const selected = yield select(path(['navigation', 'criteria', 'selected']));

    if (selected) {
      const deletedCriteriaState = yield all(map(id => call(deleteCriterion, id), selected));
      const errors = pipe(filter(propEq('isSuccess', false)), map(prop('name')))(deletedCriteriaState);

      const hasError = !isNil(errors) && !isEmpty(errors);
      const hasMoreThanOneError = hasError && length(errors) > 1;
      if (hasError)
        yield put({
          type: DELETE_CRITERIA_ERROR,
          error: {
            errorMessage: `Echec de suppression d${hasMoreThanOneError ? 'es' : 'u'}  critÃ¨re${
              hasMoreThanOneError ? 's' : ''
            } ${join(',', errors)}`,
          },
        });

      const success = pipe(filter(propEq('isSuccess', true)), map(prop('id')))(deletedCriteriaState);

      if (isNil(success) || isEmpty(success)) {
        yield put({ type: RESET_PENDING_REQUEST });
        return null;
      }

      const criteria = yield select(pipe(path(['entities', 'criteria', 'byId']), values));
      const newCriteria =
        !isNil(criteria) && criteria.filter(criterion => !contains(prop('id', criterion), success));
      yield put({ type: SET_CRITERIA, payload: newCriteria });

      yield put({ type: SET_CRITERIA_SELECTED, payload: [] });
      if (!hasError) yield put({ type: DELETE_CRITERIA_SUCCESS });
    }
  } catch (error) {
    yield put({
      type: DELETE_CRITERIA_ERROR,
      error: { errorMessage: prop('message', error) },
    });
  }

  yield put({ type: RESET_PENDING_REQUEST });
  return null;
}

export default [
  takeLatest(FETCH_CRITERIA, fetchCriteria),
  takeLatest(DELETE_CRITERIA_REQUEST, deleteCriteria),
];
