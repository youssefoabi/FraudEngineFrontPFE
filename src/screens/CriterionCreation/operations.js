import { call, put, takeLatest } from 'redux-saga/effects';
import { isNilOrEmpty } from 'ramda-adjunct';

import {
  CREATE_CRITERION_ERROR,
  CREATE_CRITERION_REQUEST,
  CREATE_CRITERION_SUCCESS,
} from '../../components/CriterionForm/actions';
import { PENDING_REQUEST, RESET_PENDING_REQUEST } from '../../components/ServerActionBackdrop/actions';
import clientApi from '../../client-api';
import parse from './parse';

export function* createCriterion({ payload }) {
  yield put({ type: PENDING_REQUEST });

  try {
    const { criterion } = payload;
    if (isNilOrEmpty(criterion)) return null;

    yield call(clientApi().createCriterion, parse(criterion));

    yield put({ type: RESET_PENDING_REQUEST });
    yield put({ type: CREATE_CRITERION_SUCCESS });
  } catch (error) {
    yield put({ type: RESET_PENDING_REQUEST });
    yield put({
      type: CREATE_CRITERION_ERROR,
      error: { errorMessage: "Erreur d'enregistrement du crit√®re" },
    });
  }

  return true;
}

export default [takeLatest(CREATE_CRITERION_REQUEST, createCriterion)];
