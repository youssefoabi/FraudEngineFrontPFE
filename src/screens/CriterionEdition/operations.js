import { call, put, takeLatest } from 'redux-saga/effects';

import {
  EDIT_CRITERION_ERROR,
  EDIT_CRITERION_REQUEST,
  EDIT_CRITERION_SUCCESS,
} from '../../components/CriterionForm/actions';
import { PENDING_REQUEST, RESET_PENDING_REQUEST } from '../../components/ServerActionBackdrop/actions';
import clientApi from '../../client-api';
import parse from '../CriterionCreation/parse';

export function* saveCriterion({ payload }) {
  yield put({ type: PENDING_REQUEST });

  try {
    const { criterion } = payload;
    if (!criterion) return null;

    yield call(clientApi().updateCriterion, parse(criterion));
    yield put({ type: RESET_PENDING_REQUEST });
    yield put({ type: EDIT_CRITERION_SUCCESS });
  } catch (error) {
    yield put({ type: RESET_PENDING_REQUEST });
    yield put({
      type: EDIT_CRITERION_ERROR,
      error: { errorMessage: "Erreur d'enregistrement de moteur de recherche" },
    });
  }

  return true;
}

export default [takeLatest(EDIT_CRITERION_REQUEST, saveCriterion)];
