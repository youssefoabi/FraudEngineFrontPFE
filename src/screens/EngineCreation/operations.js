import { call, put, takeLatest } from 'redux-saga/effects';
import { isNilOrEmpty } from 'ramda-adjunct';

import {
  CREATE_ENGINE_ERROR,
  CREATE_ENGINE_REQUEST,
  CREATE_ENGINE_SUCCESS,
} from '../../components/EngineForm/actions';
import { PENDING_REQUEST, RESET_PENDING_REQUEST } from '../../components/ServerActionBackdrop/actions';
import clientApi from '../../client-api';
import parse from './parse';

export function* createEngine({ payload }) {
  yield put({ type: PENDING_REQUEST });

  try {
    const { engine } = payload;
    if (isNilOrEmpty(engine)) return null;

    yield call(clientApi().createEngine, parse(engine));

    yield put({ type: RESET_PENDING_REQUEST });
    yield put({ type: CREATE_ENGINE_SUCCESS });
  } catch (error) {
    yield put({ type: RESET_PENDING_REQUEST });
    yield put({
      type: CREATE_ENGINE_ERROR,
      error: { errorMessage: "Erreur d'enregistrement du moteur de recherche" },
    });
  }

  return true;
}

export default [takeLatest(CREATE_ENGINE_REQUEST, createEngine)];
