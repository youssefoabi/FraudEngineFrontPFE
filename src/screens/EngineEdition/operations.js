import { call, put, takeLatest } from 'redux-saga/effects';

import {
  EDIT_ENGINE_ERROR,
  EDIT_ENGINE_REQUEST,
  EDIT_ENGINE_SUCCESS,
} from '../../components/EngineForm/actions';
import { PENDING_REQUEST, RESET_PENDING_REQUEST } from '../../components/ServerActionBackdrop/actions';
import clientApi from '../../client-api';
import parse from '../EngineCreation/parse';

export function* saveEngine({ payload }) {
  yield put({ type: PENDING_REQUEST });

  try {
    const { engine } = payload;
    if (!engine) return null;

    yield call(clientApi().updateEngine, parse(engine));
    yield put({ type: RESET_PENDING_REQUEST });
    yield put({ type: EDIT_ENGINE_SUCCESS });
  } catch (error) {
    yield put({ type: RESET_PENDING_REQUEST });
    yield put({
      type: EDIT_ENGINE_ERROR,
      error: { errorMessage: "Erreur d'enregistrement de moteur de recherche" },
    });
  }

  return true;
}

export default [takeLatest(EDIT_ENGINE_REQUEST, saveEngine)];
