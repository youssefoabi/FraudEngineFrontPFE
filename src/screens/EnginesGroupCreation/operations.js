import { call, put, takeLatest } from 'redux-saga/effects';
import { isEmpty, isNil } from 'ramda';

import {
  CREATE_ENGINESGROUP_ERROR,
  CREATE_ENGINESGROUP_REQUEST,
  CREATE_ENGINESGROUP_SUCCESS,
} from '../../components/EnginesGroupForm/actions';
import { PENDING_REQUEST, RESET_PENDING_REQUEST } from '../../components/ServerActionBackdrop/actions';
import clientApi from '../../client-api';
import parse from './parse';

export function* createEnginesGroup({ payload }) {
  yield put({ type: PENDING_REQUEST });

  try {
    const { enginesGroup } = payload;
    if (isNil(enginesGroup) || isEmpty(enginesGroup)) return null;

    yield call(clientApi().createEnginesGroup, parse(enginesGroup));

    yield put({ type: RESET_PENDING_REQUEST });
    yield put({ type: CREATE_ENGINESGROUP_SUCCESS });
  } catch (error) {
    yield put({ type: RESET_PENDING_REQUEST });
    yield put({
      type: CREATE_ENGINESGROUP_ERROR,
      error: { errorMessage: "Erreur d'enregistrement du groupe de moteurs" },
    });
  }

  return true;
}

export default [takeLatest(CREATE_ENGINESGROUP_REQUEST, createEnginesGroup)];
