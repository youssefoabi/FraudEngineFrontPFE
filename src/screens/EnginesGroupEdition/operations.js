import { call, put, takeLatest } from 'redux-saga/effects';

import {
  EDIT_ENGINESGROUP_ERROR,
  EDIT_ENGINESGROUP_REQUEST,
  EDIT_ENGINESGROUP_SUCCESS,
} from '../../components/EnginesGroupForm/actions';
import { PENDING_REQUEST, RESET_PENDING_REQUEST } from '../../components/ServerActionBackdrop/actions';
import clientApi from '../../client-api';
import parse from '../EnginesGroupCreation/parse';

export function* saveEnginesGroup({ payload }) {
  yield put({ type: PENDING_REQUEST });

  try {
    const { enginesGroup } = payload;
    if (!enginesGroup) return null;
    yield call(clientApi().updateEnginesGroup, parse(enginesGroup));

    yield put({ type: RESET_PENDING_REQUEST });
    yield put({ type: EDIT_ENGINESGROUP_SUCCESS });
  } catch (error) {
    yield put({ type: RESET_PENDING_REQUEST });
    yield put({
      type: EDIT_ENGINESGROUP_ERROR,
      error: { errorMessage: "Erreur d'enregistrement du groupe de moteurs" },
    });
  }

  return true;
}

export default [takeLatest(EDIT_ENGINESGROUP_REQUEST, saveEnginesGroup)];
