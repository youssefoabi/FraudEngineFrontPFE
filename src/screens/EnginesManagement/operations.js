import { all, call, put, select, takeLatest } from 'redux-saga/effects';
import {
  contains,
  filter,
  isEmpty,
  isNil,
  join,
  length,
  map,
  path,
  pathOr,
  pipe,
  prop,
  propEq,
  values,
} from 'ramda';

import {
  DELETE_ENGINES_ERROR,
  DELETE_ENGINES_REQUEST,
  DELETE_ENGINES_SUCCESS,
  FETCH_ENGINES,
  FETCH_ENGINES_ERROR,
  SET_ENGINES,
  SET_ENGINES_SELECTED,
} from '../../components/EnginesDataGrid/actions';
import { PENDING_REQUEST, RESET_PENDING_REQUEST } from '../../components/ServerActionBackdrop/actions';
import clientApi from '../../client-api';

export const parseResult = pathOr([], ['data']);

export function* fetchEngines() {
  try {
    const result = yield call(clientApi().getEngines);

    const engines = parseResult(result);

    yield put({ type: SET_ENGINES, payload: engines });

    yield put({ type: SET_ENGINES_SELECTED, payload: [] });
  } catch (error) {
    yield put({ type: FETCH_ENGINES_ERROR, error });
  }
}

export function* deleteEngine(id) {
  try {
    yield call(clientApi().deleteEngine, id);
  } catch (error) {
    const name = yield select(path(['entities', 'engines', 'byId', id, 'name']));
    return { isSuccess: false, name, id };
  }

  return { isSuccess: true, id };
}

export function* deleteEngines() {
  yield put({ type: PENDING_REQUEST });

  try {
    const selected = yield select(path(['navigation', 'engines', 'selected']));

    if (selected) {
      const deletedEnginesState = yield all(map(id => call(deleteEngine, id), selected));
      const errors = pipe(filter(propEq('isSuccess', false)), map(prop('name')))(deletedEnginesState);

      const hasError = !isNil(errors) && !isEmpty(errors);
      const hasMoreThanOneError = hasError && length(errors) > 1;
      if (hasError)
        yield put({
          type: DELETE_ENGINES_ERROR,
          error: {
            errorMessage: `Echec de suppression d${hasMoreThanOneError ? 'es' : 'u'}  moteur${
              hasMoreThanOneError ? 's' : ''
            } ${join(',', errors)}`,
          },
        });

      const success = pipe(filter(propEq('isSuccess', true)), map(prop('id')))(deletedEnginesState);

      if (isNil(success) || isEmpty(success)) {
        yield put({ type: RESET_PENDING_REQUEST });
        return null;
      }

      const engines = yield select(pipe(path(['entities', 'engines', 'byId']), values));
      const newEngines = !isNil(engines) && engines.filter(engine => !contains(prop('id', engine), success));
      yield put({ type: SET_ENGINES, payload: newEngines });

      yield put({ type: SET_ENGINES_SELECTED, payload: [] });
      if (!hasError) yield put({ type: DELETE_ENGINES_SUCCESS });
    }
  } catch (error) {
    yield put({
      type: DELETE_ENGINES_ERROR,
      error: { errorMessage: prop('message', error) },
    });
  }

  yield put({ type: RESET_PENDING_REQUEST });
  return null;
}

export default [takeLatest(FETCH_ENGINES, fetchEngines), takeLatest(DELETE_ENGINES_REQUEST, deleteEngines)];
