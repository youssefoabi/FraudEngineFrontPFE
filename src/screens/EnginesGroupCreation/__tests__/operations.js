import { itShouldMatchSequence, NO_PARAM } from '@cdiscount/jest-utils';

import { createEnginesGroup } from '../operations';
import enginesGroupFixture from '../../EnginesGroupEdition/__fixtures__';

jest.mock('react-global-configuration');

describe('EnginesGroupCreation operations', () => {
  describe('saga', () => {
    const enginesGroup = { payload: { enginesGroup: enginesGroupFixture } };
    const nullEnginesGroup = { payload: { enginesGroup: null } };

    itShouldMatchSequence(
      'createEnginesGroup with null enginesGroup',
      () => createEnginesGroup(nullEnginesGroup),
      [NO_PARAM, NO_PARAM, NO_PARAM],
    );
    itShouldMatchSequence('createEnginesGroup with success', () => createEnginesGroup(enginesGroup), [
      NO_PARAM,
      NO_PARAM,
      NO_PARAM,
    ]);
    itShouldMatchSequence('createEnginesGroup with failure', () => createEnginesGroup(enginesGroup), [
      NO_PARAM,
      NO_PARAM,
      new Error(),
    ]);
  });
});
