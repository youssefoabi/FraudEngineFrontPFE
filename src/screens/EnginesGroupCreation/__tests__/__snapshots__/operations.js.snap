// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnginesGroupCreation operations saga createEnginesGroup with failure 1`] = `
Array [
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "PENDING_REQUEST",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "engines": Array [
              Object {
                "engineId": 1,
                "engineLabel": "engineLabel",
                "engineName": "engine 1",
                "isEnable": true,
                "order": 1,
                "pictogramUrl": "pictogramUrl",
              },
              Object {
                "engineId": 2,
                "engineLabel": "engineLabel 2",
                "engineName": "engine 2",
                "isEnable": false,
                "order": 0,
                "pictogramUrl": "pictogramUrl 2",
              },
            ],
            "groupCode": 2,
            "groupId": 1,
            "groupName": "groupName",
            "isEnable": true,
          },
        ],
        "context": null,
        "fn": [Function],
      },
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "RESET_PENDING_REQUEST",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": Object {
            "errorMessage": "Erreur d'enregistrement du groupe de moteurs",
          },
          "type": "CREATE_ENGINESGROUP_ERROR",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": true,
    "value": true,
  },
]
`;

exports[`EnginesGroupCreation operations saga createEnginesGroup with null enginesGroup 1`] = `
Array [
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "PENDING_REQUEST",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": true,
    "value": null,
  },
]
`;

exports[`EnginesGroupCreation operations saga createEnginesGroup with success 1`] = `
Array [
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "PENDING_REQUEST",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "engines": Array [
              Object {
                "engineId": 1,
                "engineLabel": "engineLabel",
                "engineName": "engine 1",
                "isEnable": true,
                "order": 1,
                "pictogramUrl": "pictogramUrl",
              },
              Object {
                "engineId": 2,
                "engineLabel": "engineLabel 2",
                "engineName": "engine 2",
                "isEnable": false,
                "order": 0,
                "pictogramUrl": "pictogramUrl 2",
              },
            ],
            "groupCode": 2,
            "groupId": 1,
            "groupName": "groupName",
            "isEnable": true,
          },
        ],
        "context": null,
        "fn": [Function],
      },
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "RESET_PENDING_REQUEST",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "CREATE_ENGINESGROUP_SUCCESS",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": true,
    "value": true,
  },
]
`;
